Build first draft of indexer and planner

dunkelziffer opened the pull request:
This contains drafts for both the "indexer" and the "planner" (which is currently also taking care of the file copying (= "plan execution")).

I consider this still a "spiking phase", so there are no tests yet, because I'm still constantly shuffling stuff around and changing interfaces. I'll write tests as soon as stuff has settled into a more stable state, pinky promise :sweat_smile: 

The code is also very often pretty ugly. I'd still consider this mergeable so you can hook it into the CLI and actually get to a first working version.

The current state: 
- Assumption: the target repo doesn't have any component files yet
- This call copies all component pack files to the project:
```ruby
WeirdPhlex::Core::Planner.initial_full_installation_plan.perform!
```

How to use this currently:
- Have the target project and component pack dev repos checked out at `avi-sample`
- Run the code above from the target project Rails console

dunkelziffer commented:
I consider this still a "spiking phase", so there are no tests yet, because I'm still constantly shuffling stuff around and changing interfaces. I'll write tests as soon as stuff has settled into a more stable state, pinky promise :sweat_smile: 

The code is also very often pretty ugly. I'd still consider this mergeable so you can hook it into the CLI and actually get to a first working version.

The current state: 
- Assumption: the target repo doesn't have any component files yet
- This call copies all component pack files to the project:
```ruby
WeirdPhlex::Core::Planner.initial_full_installation_plan.perform!
```

N3BCKN commented on lib/weird_phlex/core/component_pack/variant.rb:
```diff
@@ -0,0 +1,69 @@
+        end
+
+        def gem_path(name)
+          path, error, status = Open3.capture3("bundle", "show", name)
```
How about something like: `Bundler.rubygems.find_name(name).first.full_gem_path` instead?

dunkelziffer replied:
```ruby
Gem.loaded_specs[name].gem_dir
```
:smile:

N3BCKN reviewed (commented):
all right, few improvements to increase basic readability would be really welcomed here. 

I could not test it that much, but the method `component_packs` from Indexer you've mentioned [here](https://github.com/weird-phlex/weird_phlex/pull/3) didn't work at all. It's non existent at this stage.

dunkelziffer commented on lib/weird_phlex/core/project/component.rb:
```diff
@@ -0,0 +1,31 @@
+
+module WeirdPhlex
+  module Core
+    module Project
```
It's "in contrast to `ComponentPack`". I though `Project` would be sufficient. I don't think `TargetProject` or anything even more specific will be necessary. We can change it later, if necessary.

dunkelziffer commented on lib/weird_phlex/core/planner/plan.rb:
```diff
@@ -0,0 +1,43 @@
+        end
+
+        def copy(source_file, target_file)
+          add(CopyAction.new(source_file, target_file))
```
I very rarely use dependency injection in Ruby/Rails. The benefit for testing is very minimal with RSpec. And I'm not sure we ever need the flexibility and actually want to swap this out.

dunkelziffer commented on lib/weird_phlex/core/project/file.rb:
```diff
@@ -0,0 +1,143 @@
+          @project_path = project_path
+          @relative_project_path = @path.to_s.delete_prefix("#{@project_path}/")
+
+          @data = weird_phlex_hash
```
This will surely happen in the upcoming refactorings. But I want to have a better overview of the data flow and attributes in the initializer for now. If I hide and wrap everything in methods, it's harder to move stuff around.

dunkelziffer commented:
> I could not test it that much, but the method `component_packs` from Indexer you've mentioned [here](https://github.com/weird-phlex/weird_phlex/pull/3) didn't work at all. It's non existent at this stage.

That comment was the original API from when I first requested a review. It was outdated. I moved the contents of my up-to-date comment into the PR description and deleted the old description.
